### ISSUE TITLE
match_single_binding is false positive on temporaries 

### ISSUE THREAD

**Issue by MikailBag** [rust-lang/rust-clippy#6614]
Lint name: `match_single_binding`

I tried this code:

```rust
fn main() {
    match format_args!("hello {}", "world") {
        f => dbg!(f),
    };
}
```
([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9573dc18836305f5f860b3dc9b7d875a))

Clippy prints a warning:
```
warning: this match could be written as a `let` statement
```
With a proposal:
```
let f = format_args!("hello {}", "world");
dbg!(f);
```
But suggested does not work, because `f` is bound to a temporary value and its lifetime ends before `dbg!`. 
### Meta

- `cargo clippy -V`: 0.1.51 (2021-01-20 a4cbb44)
- `rustc -Vv`: 

```
rustc 1.49.0 (e1884a8e3 2020-12-29)
binary: rustc
commit-hash: e1884a8e3c3e813aada8254edfa120e85bf5ffca
commit-date: 2020-12-29
host: x86_64-unknown-linux-gnu
release: 1.49.0
```

**Comment by camsteffen** [comment]
I tried to fix this with `HIR` and `ScopeTree` to no avail. I think this will require MIR analysis. Of course we could special-case `format_args!` but I think that is too rare of a use-case for that.

**Comment by Fishrock123** [comment]
Just ran into this with `todo!()`, which seems very silly:

```
warning: this match could be replaced by its body itself
  --> src/thing.rs:19:9
   |
19 | /         match event.action() {
20 | |             _ => todo!("make thing do"),
21 | |         }
   | |_________^ help: consider using the match body instead: `todo!("make thing do")`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_single_binding
```

**Comment by camsteffen** [comment]
I don't think it would be worth it to special-case `todo!` since that indicates incomplete code by definition.

**Comment by Fishrock123** [comment]
Oh, on re-reading this maybe this is a separate issue. I think it _should not_ warn in the case of `todo!`. I think it would be worth it, that should be highlighted in different ways.

**Comment by bgeron** [comment]
I just ran into the same issue.

To be honest — matching with a single match arm _is_ a bit silly, `let` is normally meant for that. And if you put a block in the match, now you have two indentation levels!

It seems nice to have different syntax for this use case. For instance:

```rs
let f = format_args!("hello {}", "world") in {
    dbg!(f),
    dbg!(f),
};
```

But the lint needs a fix irrespective of any new syntax.
